openerp/addons/base/res/res_currency.py

```diff
@@ -1,134 +1,123 @@
     _description = "Currency"
-    _columns = {
-        # Note: 'code' column was removed as of v6.0, the 'name' should now hold the ISO code.
-        'name': fields.char('Currency', size=3, required=True, help="Currency Code (ISO 4217)"),
-        'symbol': fields.char('Symbol', size=4, help="Currency sign, to be used when printing amounts."),
-        'rate': fields.function(_get_current_rate, string='Current Rate', digits=(12,6),
-            help='The rate of the currency to the currency of rate 1.'),
-        'rate_ids': fields.one2many('res.currency.rate', 'currency_id', 'Rates'),
-        'rounding': fields.float('Rounding Factor', digits=(12,6)),
-        'decimal_places': fields.function(_decimal_places, string='Decimal Places', type='integer'),
-        'active': fields.boolean('Active'),
-        'position': fields.selection([('after','After Amount'),('before','Before Amount')], 'Symbol Position', help="Determines where the currency symbol should be placed after or before the amount.")
-    }
-    _defaults = {
-        'active': 1,
-        'position' : 'after',
-        'rounding': 0.01,
-    }
+    _order = "name"
+
+    # Note: 'code' column was removed as of v6.0, the 'name' should now hold the ISO code.
+    name = fields.Char(string='Currency', size=3, required=True, help="Currency Code (ISO 4217)")
+    symbol = fields.Char(help="Currency sign, to be used when printing amounts.")
+    rate = fields.Float(compute='_compute_current_rate', string='Current Rate', digits=(12, 6),
+                        help='The rate of the currency to the currency of rate 1.')
+    rate_ids = fields.One2many('res.currency.rate', 'currency_id', string='Rates')
+    rounding = fields.Float(string='Rounding Factor', digits=(12, 6), default=0.01)
+    decimal_places = fields.Integer(compute='_compute_decimal_places')
+    active = fields.Boolean(default=True)
+    position = fields.Selection([('after', 'After Amount'), ('before', 'Before Amount')], default='after',
+        string='Symbol Position', help="Determines where the currency symbol should be placed after or before the amount.")
+    date = fields.Date(compute='_compute_date')
+
     _sql_constraints = [
         ('unique_name', 'unique (name)', 'The currency code must be unique!'),
     ]
-    _order = "name"
 
-    date = fields2.Date(compute='compute_date')
+    @api.multi
+    def _compute_current_rate(self):
+        date = self.env.context.get('date') or fields.Datetime.now()
+        company_id = self.env.context.get('company_id', self.env.user._get_company().id)
+        self.env.cr.execute("""SELECT currency_id, rate FROM res_currency_rate
+                           WHERE currency_id in %s
+                             AND name <= %s
+                             AND (company_id is null
+                                 OR company_id = %s)
+                        ORDER BY company_id, name""",
+                                (tuple(self.ids,), date, company_id))
```

Comment:
Ouch! This query is inefficient: it retrieves **all** rates for the given currencies up to the given date, while only the last ones are useful for the computation... The former query only retrieved one row per currency.